### PR TITLE
Display popular locations

### PR DESCRIPTION
**Pull request by theresenieva**
Display popular locations on the page.

I've created a servlet to return a JSON array of the popular locations and JavaScript code to display those locations. I chose to display the popular locations as a list below the map when the user clicks a button. I think this would be cleaner than having the list update whenever a new location is added.

### CODE REVIEW

File: GoMeet/src/main/java/com/google/sps/servlets/PopularLocationServlet.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    String json = ServletUtil.convertToJson(popularLocations);
+    response.getWriter().println(json);
+  }
+
+  /** Returns a list of the locations with the highest vote count. */
+  private List<Location> getPopularLocation() {
+    List<Location> popularLocations = new ArrayList<>();
+    List<Location> allLocations = locationDao.getAll();
+
+    int maxVoteCount = 0;
```
Comment (theresenieva) [inline GoMeet/src/main/java/com/google/sps/servlets/PopularLocationServlet.java:46]: I decided to iterate through the list of locations to find the most popular locations as it was:
(1) O(n) each location is visited once.
(2) Given time-constraints, this approach is simple to implement,
(3) In our design doc, we specified that each meeting will have max. 5 locations (so a small set of data).

I considered sorting the list first based on the vote count and then returning the first elements in the list but Collection.sort() has a complexity of O(n*log(n)).

In future, if this feature were to be scaled up, I think storing a list of popular locations as a property of the meeting entity would be more efficient (and updating it when a new location is added) as each location entity for that meeting does not need to be read when requesting the popular locations.

File: GoMeet/src/main/java/com/google/sps/dao/LocationDao.java [hunk 2]
```diff
@@ -8,6 +8,8 @@
 import com.google.appengine.api.datastore.KeyFactory;
 import com.google.appengine.api.datastore.PreparedQuery;
 import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
```
Comment (vivianho18) [inline GoMeet/src/main/java/com/google/sps/dao/LocationDao.java]: Why did you add these two imports?

File: GoMeet/src/main/java/com/google/sps/servlets/PopularLocationServlet.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        continue;
+      }
+      if (currVoteCount > maxVoteCount) {
+        popularLocations.clear();
+        maxVoteCount = currVoteCount;
+      }
+      popularLocations.add(location);
+    }
+    return popularLocations;
+  }
```
Comment (vivianho18) [inline GoMeet/src/main/java/com/google/sps/servlets/PopularLocationServlet.java:59]: Kind of confused by how this function works. Why do you clear the whole popularLocations arraylist each time the current iteration has a higher vote count? How many locations do you expect to be included in the final result of popularLocations (e.g. top 3 most popular locations)? Might be good to explain this in a comment above the function.

File: GoMeet/src/main/webapp/js/meeting-location-dao.js [hunk 4]
```diff
@@ -13,6 +15,16 @@ class MeetingLocationDAO {
     return locations;
   }
 
+   /**
+   * Fetches the popular location data from the servlet.
+   * Returns a JSON array of the location data.
+   */
+  static async fetchPopularLocations() {
+    let response = await fetch(this.popularEndPoint);
+    let locations = await response.json();
```
Comment (vivianho18) [inline GoMeet/src/main/webapp/js/meeting-location-dao.js]: nit: chain this with arrow functions

File: GoMeet/src/main/webapp/meeting-event.html [hunk 5]
```diff
@@ -21,6 +21,9 @@ <h3>Meeting times here</h3>
       <h2>Meeting Locations</h2>
       <p>Click on the map to propose a new location.</p>
       <div id="map"></div></br>
+      <button onclick="displayPopularLocations()">Most Popular Locations</button><br>
+      <p>Popular Locations:</p>
```
Comment (vivianho18) [inline GoMeet/src/main/webapp/meeting-event.html:25]: Should this be displayed regardless of whether the user has clicked the button above? Maybe reformat to be something more like: 
```
<p>Popular locations:</p> 
<button onclick="displayPopularLocations()">Click to see the most popular locations</button><br>
<ul id="popular-locations-container"></ul> 
```

File: GoMeet/src/main/webapp/test/spec/spec-interactive-map.js [hunk 6]
```diff
@@ -144,3 +144,38 @@ describe ('Build Info Window Vote', function() {
...
+
+  it ('Should add 3 list elements to the popular location list',
+      async function() {     
+    spyOn(MeetingLocationDAO, 'fetchPopularLocations').and.returnValue(LOCATIONS);
+
+    await displayPopularLocations();
+
+    let listSize = 
+        document.getElementById('popular-locations-container').childNodes.length;
+    expect(listSize).toBe(3);
```
Comment (vivianho18) [inline GoMeet/src/main/webapp/test/spec/spec-interactive-map.js:168]: Could add some more expects for checking the actual list elements

File: GoMeet/src/main/java/com/google/sps/dao/LocationDao.java (hunk 2 above)
Comment (theresenieva) [reply GoMeet/src/main/java/com/google/sps/dao/LocationDao.java]: Oops, I think those two imports found their way in from another feature. I started working on it on this branch and then decided to move that other feature to a new branch. Will remove to make things clearer :)

File: GoMeet/src/main/webapp/test/spec/spec-interactive-map.js [hunk 7]
```diff
@@ -144,3 +144,38 @@ describe ('Build Info Window Vote', function() {
     expect(noteText).toBe(NOTE_A);
   });
 });
+
+/** Tests for displaying popular locations. */
+describe ('Display Popular Location', function() {
```
Comment (vivianho18) [inline GoMeet/src/main/webapp/test/spec/spec-interactive-map.js:149]: Does the popular location list order the locations by number of votes? i.e. most popular location is number 1 on the list - or is it just top 3 in any order? Could be worth adding another test cases with vote count 5, 4, 3 and check the order of the locations in the list is as expected. Also, what happens if one of the locations has 0 votes but there are only 3 locations in total? Should/will it be included in the popular list?

File: GoMeet/src/test/java/com/google/sps/PopularLocationServletTest.java [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+
```
Comment (vivianho18) [inline GoMeet/src/test/java/com/google/sps/PopularLocationServletTest.java]: nit - remove this extra newline

File: GoMeet/src/test/java/com/google/sps/PopularLocationServletTest.java [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+/** Tests for PopularLocationServlet.java */
+@RunWith(JUnit4.class)
+public class PopularLocationServletTest {
+  // Values to use for testing
+  private final Location LOCATION_A =
+      new Location("Burger Bonanza", 15.0, 150.0, "I like burgers!", 1);
+  private final Location LOCATION_B =
+      new Location("Pastry Patisserie", 20.0, 10.0, "Pastries are yum!", 5);
+  private final Location LOCATION_C = new Location("Pizza Place", 20.0, 10.0, "Nice Cheese!", 5);
```
Comment (vivianho18) [inline GoMeet/src/test/java/com/google/sps/PopularLocationServletTest.java]: nit: newline after "new" so that it matches with the format of location A and B

File: GoMeet/src/test/java/com/google/sps/PopularLocationServletTest.java [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+  public void setUp() throws IOException {
+    request = mock(HttpServletRequest.class);       
+    response = mock(HttpServletResponse.class);  
+    mockedLocationDao = mock(LocationDao.class);
+    when(response.getWriter()).thenReturn(writer);
+  }
+
+  @After
+  public void tearDown() {
+  }
```
Comment (vivianho18) [inline GoMeet/src/test/java/com/google/sps/PopularLocationServletTest.java:58]: Can this be deleted?

File: GoMeet/src/test/java/com/google/sps/PopularLocationServletTest.java [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+    servlet.doGet(request, response);
+
+    // Check if the popular location was sent in the response
+    stringWriter.flush();
+    Type locationListType = new TypeToken<ArrayList<Location>>(){}.getType();
+    ArrayList<Location> locationList = gson.fromJson(stringWriter.toString(), locationListType);
+
+    assertTrue(locationList.contains(LOCATION_B));
+    assertFalse(locationList.contains(LOCATION_A));
+  }
```
Comment (vivianho18) [inline GoMeet/src/test/java/com/google/sps/PopularLocationServletTest.java:82]: nit - also assert size of locationList is 1

File: GoMeet/src/test/java/com/google/sps/PopularLocationServletTest.java [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+
+    // Check if the popular locations were sent in the response
+    stringWriter.flush();
+    Type locationListType = new TypeToken<ArrayList<Location>>(){}.getType();
+    ArrayList<Location> locationList = gson.fromJson(stringWriter.toString(), locationListType);
+
+    assertTrue(locationList.contains(LOCATION_B));
+    assertTrue(locationList.contains(LOCATION_C));
+    assertFalse(locationList.contains(LOCATION_A));   
+  }
```
Comment (vivianho18) [inline GoMeet/src/test/java/com/google/sps/PopularLocationServletTest.java:108]: nit - same as above, ensure list size is 2

File: GoMeet/src/test/java/com/google/sps/PopularLocationServletTest.java [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+    when(mockedLocationDao.getAll()).thenReturn(listToReturn);
+
+    PopularLocationServlet servlet = new PopularLocationServlet();
+    servlet.setDao(mockedLocationDao);
+    servlet.doGet(request, response);
+
+    // Check if the popular location was sent in the response
+    stringWriter.flush();
+    Type locationListType = new TypeToken<ArrayList<Location>>(){}.getType();
+    ArrayList<Location> locationList = gson.fromJson(stringWriter.toString(), locationListType);
```
Comment (vivianho18) [inline GoMeet/src/test/java/com/google/sps/PopularLocationServletTest.java]: optional - can extract these lines into a helper function to avoid repeated code

File: GoMeet/src/main/java/com/google/sps/servlets/PopularLocationServlet.java (hunk 3 above)
Comment (theresenieva) [reply GoMeet/src/main/java/com/google/sps/servlets/PopularLocationServlet.java:59]: So for this function, if the locations stored are: 
- Location A, VoteCount: 5
- Location B, VoteCount: 5
- Location C, VoteCount: 1
Then Location A and Location B are returned in the list. 

There is no expected number of popular locations. It returns the locations in the database with the highest vote count. This can be multiple locations. 

I'll update the Javadoc to make it clearer.

**Review by vivianho18 — COMMENTED** [review]
A few suggestions and clarifications but other than that LGTM :)

File: GoMeet/src/main/webapp/meeting-event.html (hunk 5 above)
Comment (theresenieva) [reply GoMeet/src/main/webapp/meeting-event.html:25]: "Popular Locations" is displayed regardless of whether the button is clicked. It is below the button. I thought it looked cleaner having the popular locations heading right above the list. I'm uploading a screenshot but it is taking a while haha.

File: GoMeet/src/main/webapp/test/spec/spec-interactive-map.js (hunk 7 above)
Comment (theresenieva) [reply GoMeet/src/main/webapp/test/spec/spec-interactive-map.js:149]: Covered in previous comment but only the locations with the high votes are listed (in no particular order).

**Comment by theresenieva** [comment]
Screen Shot 2021-02-09 at 8 59 30 am

File: GoMeet/src/test/java/com/google/sps/PopularLocationServletTest.java (hunk 13 above)
Comment (theresenieva) [reply GoMeet/src/test/java/com/google/sps/PopularLocationServletTest.java]: Yup, I've decided to set up the servlet with the mockedDao is the @Before setup to remove the repeated code :)

File: GoMeet/src/main/java/com/google/sps/servlets/PopularLocationServlet.java (hunk 3 above)
Comment (vivianho18) [reply GoMeet/src/main/java/com/google/sps/servlets/PopularLocationServlet.java:59]: Ok that makes sense

File: GoMeet/src/main/webapp/js/interactive-map.js [hunk 14]
```diff
@@ -137,8 +137,16 @@ function createPopularLocationElement(location) {
   const liElement = document.createElement('li');
   liElement.className = 'location';
 
-  liElement.append('Title: ', location.title, document.createElement('br'),
-      'Number of Votes: ', location.voteCount);
+  let titleContainer = document.createElement('span');
```
Comment (theresenieva) [inline GoMeet/src/main/webapp/js/interactive-map.js:140]: Note: there is some repeated code here. 
I've made a helper function for adding an id to a span element with given innertext in #45. I plan to:
1) Merge this branch with master.
2) Rebase #45 onto master and update this function to use the helper function.
3) Merge that brach with master.

**Review by anna-giang — APPROVED** [review]
LGTM, with the changes commented 😊 Great suggestions @vivianho18

File: GoMeet/src/main/java/com/google/sps/servlets/UpdateLocationServlet.java [hunk 15]
```diff
@@ -1,21 +1,20 @@
...
-import com.google.sps.data.ServletUtil;
-import com.google.sps.data.ErrorMessages;
-
 import com.google.gson.Gson;
-import java.io.IOException;
+import com.google.sps.data.ErrorMessages;
+import com.google.sps.data.ServletUtil;
 import java.util.ArrayList;
 import java.util.Collection;
+import java.io.IOException;
```
Comment (BipBopBoop) [inline GoMeet/src/main/java/com/google/sps/servlets/UpdateLocationServlet.java]: nit: Generally nice to have these in alphabetical order

File: GoMeet/src/main/java/com/google/sps/servlets/UpdateLocationServlet.java [hunk 16]
```diff
@@ -1,21 +1,20 @@
...
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import main.java.com.google.sps.data.Location;
+import main.java.com.google.sps.dao.Dao;
+import main.java.com.google.sps.dao.LocationDao;
+import main.java.com.google.sps.data.Location;
```
Comment (BipBopBoop) [inline GoMeet/src/main/java/com/google/sps/servlets/UpdateLocationServlet.java:16]: nit: same as above.

File: GoMeet/src/main/webapp/js/interactive-map.js (hunk 14 above)
Comment (BipBopBoop) [reply GoMeet/src/main/webapp/js/interactive-map.js:140]: sg :)

File: GoMeet/src/main/webapp/js/interactive-map.js [hunk 17]
```diff
@@ -131,3 +131,37 @@ function buildInfoWindowVote(title, voteCount, note, keyString) {
...
+  return liElement;
+}
+
+/** Fetches the popular location data and adds it to the DOM. */
+async function displayPopularLocations() {
+  const popularLocationElement = 
+      document.getElementById('popular-locations-container');
+  popularLocationElement.innerHTML = '';
+
+  let json = await MeetingLocationDAO.fetchPopularLocations();
```
Comment (BipBopBoop) [inline GoMeet/src/main/webapp/js/interactive-map.js]: What happens if this fails? Do we need to handle that kind of error?

File: GoMeet/src/main/webapp/test/SpecRunner.html [hunk 18]
```diff
@@ -50,6 +50,7 @@
       <p>Votes table</p>
   </div> 
   <div id="map"></div></br>
+  <ul style="display: none;" id="popular-locations-container"></ul>
```
Comment (BipBopBoop) [inline GoMeet/src/main/webapp/test/SpecRunner.html:53]: Nit: It's not ideal to be putting in-line styling in HTML (https://google.github.io/styleguide/htmlcssguide.html#Separation_of_Concerns). But seeing as though it's already being done elsewhere in the file, I guess it's okay to leave it.

File: GoMeet/src/main/webapp/test/spec/spec-interactive-map.js [hunk 19]
```diff
@@ -144,3 +144,67 @@ describe ('Build Info Window Vote', function() {
...
+  const LOCATIONS =
```
Comment (BipBopBoop) [inline GoMeet/src/main/webapp/test/spec/spec-interactive-map.js:150]: You should add a location that isn't popular and doesn't have the same votecount as the others.

**Review by BipBopBoop — COMMENTED** [review]
Looks good, left some comments :)

File: GoMeet/src/main/webapp/test/SpecRunner.html (hunk 18 above)
Comment (theresenieva) [reply GoMeet/src/main/webapp/test/SpecRunner.html:53]: Ahh, I see. I'll avoid it in the future :).

File: GoMeet/src/main/webapp/test/spec/spec-interactive-map.js (hunk 19 above)
Comment (theresenieva) [reply GoMeet/src/main/webapp/test/spec/spec-interactive-map.js:150]: Just to clarify, should I add a location with a lower vote count and expect it to be displayed as well given that fetching ```fetchPopularLocations()``` will return only the most popular locations?

Comment (BipBopBoop) [reply GoMeet/src/main/webapp/test/spec/spec-interactive-map.js:150]: ahh whoops, my mistake. I misread these tests. What you have makes sense :) I forgot that we're essentially mocking out the servlet functionality and thought that we were testing it.

File: GoMeet/src/main/webapp/js/interactive-map.js (hunk 17 above)
Comment (theresenieva) [reply GoMeet/src/main/webapp/js/interactive-map.js]: Yep, I've added that in commits:
- 4e9caeb
- 14d3503

What I did was make the DAO throw an error if the locations could not be fetched. The function that calls this in ```interactive-map.js``` would catch the error and pass it to ```handleError()```, which makes an alert. I chose to use a seperate function to handle errors if in the future I want to change how errors are displayed. 

I will do the same for fetchLocations() but I thought it didn't belong in this PR. So, I will include it when I refactor the Dao :).

File: GoMeet/src/main/webapp/test/spec/spec-interactive-map.js (hunk 19 above)
Comment (theresenieva) [reply GoMeet/src/main/webapp/test/spec/spec-interactive-map.js:150]: Haha, np!

File: GoMeet/src/main/webapp/js/interactive-map.js (hunk 17 above)
Comment (BipBopBoop) [reply GoMeet/src/main/webapp/js/interactive-map.js]: SG, thanks for the explanation :)

File: GoMeet/src/main/java/com/google/sps/servlets/PopularLocationServlet.java [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+import java.util.Collections;
+import java.util.List;
+import java.util.Comparator;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import main.java.com.google.sps.dao.Dao;
+import main.java.com.google.sps.data.Location;
+import main.java.com.google.sps.dao.LocationDao;
```
Comment (BipBopBoop) [inline GoMeet/src/main/java/com/google/sps/servlets/PopularLocationServlet.java]: nit: these two should be switched, to maintain alphabetical order.

**Review by BipBopBoop — APPROVED** [review]
Thanks for the changes, just one small nit. Otherwise LGTM :)
